### PR TITLE
Implement image service to upload, list and download images

### PR DESCRIPTION
**Pull request by jbinder97**
## Image Download Service

**Uploading Images**

    Endpoint: /images/upload/
    Method: POST
    Body: JSON with a list of image URLs ({"urls": ["http://example.com/image1.jpg", ...]})

The service will download images from the provided URLs and store them.

**Getting a List of Images**

    Endpoint: /images/
    Method: GET

Returns a list of all successfully processed images.

**Retrieving an Image**

    Endpoint: /images/download/?url=<IMAGE_URL>
    Method: GET

Retrieve the image for the provided URL if it was previously stored.

## Security

- User authentication is required for all operations.
- Users can only access images they have uploaded.

## Limitations

- The service currently supports only image/jpeg and image/png formats.
- There is no error handling for failed downloads; such images are simply not stored.
- Media files are stored on device, cloud storage solutions like S3 will improve scalability and performance.

### CODE REVIEW

File: image_manager/api/views.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+        for image_url in image_urls:
+            image_url = unquote(image_url)
+
+            if Image.objects.filter(url=image_url).exists():
+                image = Image.objects.get(url=image_url)
+            else:
+                image = Image.objects.create(url=image_url, uploader=request.user)
+
+            image.download()
```
Comment (jbinder97) [inline image_manager/api/views.py:32]: The current implementation of `upload` is synchronous. If the list of URLs is long, this could lead to long processing times for the request. 

Using a task queue like Celery could be used to enable asynchronous processing.

File: image_manager/models.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+    content_type = models.CharField(max_length=20, blank=True)
+
+    created_at = models.DateTimeField(auto_now_add=True)
+
+    updated_at = models.DateTimeField(auto_now=True)
+
+    def delete(self, *args, **kwargs):
+        if self.image:
+            print(self.image.path)
```
Comment (jbinder97) [inline image_manager/models.py:38]: Remove `print(self.image.path)`

File: image_service/urls.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+router = routers.DefaultRouter()
+router.register(r'images', ImageViewSet, basename='image')
+
+urlpatterns = [
+    path('admin/', admin.site.urls),
+    path('api/', include(router.urls)),
+]
+
+if settings.DEBUG:
```
Comment (jbinder97) [inline image_service/urls.py:33]: In a production environment, we should serve media files through a web server like Nginx. Nginx is optimized for this task and can handle large numbers of simultaneous connections with a low memory footprint.

**Review by jbinder97 — COMMENTED** [review]
1. **Asynchronous Download**:
  - As discussed, implementing asynchronous processing using Celery or a similar task queue would improve performance. Each image download could be handled as a separate background task. 

2. **Error Handling**:
  - Consider adding error handling for the image download process.  You might want to log these errors or at least acknowledge failed downloads.
    
3. **Test**:
  - Review and expand our suite of unit tests. Ensure that we have comprehensive coverage for all models, views, and utility functions.
     - Focus on edge cases and error handling scenarios that might not have been fully covered.

4. **Use Nginx and Gunicorn**:
  - Replace Django's development server with Gunicorn, a production-grade WSGI server.
    - Set up Nginx as a reverse proxy in front of Gunicorn. Configure Nginx to proxy requests to Gunicorn and serve our static and media files.

5. **Dockerize the Application**:
  - Dockerizing our application will simplify deployment, ensure consistency across environments, and make it easier to scale.
